### DISCUSSION TITLE
Force-pulling files from origin while keeping new local files

### DISCUSSION THREAD

**Discussion by LO177** [community/community#22318] [New to GitHub]
<p>I would like to know if there is a cmd git command or any way to have file changes of the master repository to replace files that are different on there than local, but still keeping any new and unchanged files locally. To demonstrate:</p>
<p>[BEFORE FORCE PULL]</p>
<p>Local directory:                Master repository:</p>
<p>File_1                               File_1 (changed)</p>
<p>File_2                               File_2 (changed)</p>
<p>File_3                               File_3</p>
<p>File_4                               </p>
<p>File_5                               </p>
<p>[AFTER FORCE PULL]</p>
<p>Local directory:                Master repository:</p>
<p>File_1 (changed)              File_1 (changed)</p>
<p>File_2 (changed)              File_2 (changed)</p>
<p>File_3                               File_3</p>
<p>File_4                               </p>
<p>File_5                               </p>
<p>I want to do this before any pushing. Github desktop does not seem to like to do this, however. Is there a way I can do with a git command via command prompt (cmd)?</p>

**Comment by LO177** [comment]
<p>Anyone? If it’s not possible, it would be great if I could be at least informed of that, and in which case, what should I do? I need to get files, but at the same time I have files to upload. On SVN, I can easily pull changes to local first, why is it such a bother when using github desktop?</p>

**Comment by Wabri** [comment]
<p>Hi <a class="mention" href="/lo177">@lo177</a>,</p>
<p>Welcome to the GitHub Community forum!</p>
<p>I dont understand what do you want to do. Do you want to execute the pull before all the push? Maybe do you need an hook, you can find all the docs here -&gt; <a href="https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks" rel="nofollow noopener">Customizing-Git-Git-Hooks</a></p>
<p>Hope this helps :slightly_smiling_face:  </p>
<p>-Gabriele-</p>
<blockquote>
<p><strong>Mark helpful posts with Accept as Solution to help other users locate important info. Don’t forget to give Kudos for great content!</strong></p>
</blockquote>

**Comment by LO177** [comment]
<p>Yes, basically a pull before a push, but also without deleting files in my local directory that are on the local directory, but not on the cloud repository. Github Desktop does not let me pull without a push at all to begin with. I tried to do it via cmd instead by using <code>git pull origin master</code> command, but it still showed the same message and aborted the pull execution.</p>
<p>Is there any way to do this? Would I ideally need to use a hook to do this?</p>

**Comment by Wabri** [comment, marked as answer]
<p>Did you try using <strong>branches</strong>? This is probably the best solution for you. Create a branch called <strong>local</strong> in your local repository and keep master on remote repository. With this solution <strong>you don’t need to pull before a push</strong> every time you changes something, yuo can push every time you want to remote and from there (on github) <strong>create a pull request</strong> to see if the changes made on the local branch is able to merge in master. When you need to work in the local master you can checkout and pull.</p>
<blockquote>
<p>Github Desktop does not let me pull without a push at all to begin with. I tried to do it via cmd instead by using <code>git pull origin master</code> command, but it still showed the same message and aborted the pull execution.</p>
</blockquote>
<p>That’s because it’s not a problem of github desktop but <strong>it’s git that tell to you</strong> that it cannot merge the same branch (in this case master) that are divergent in history.</p>
<p>Assume the following history:</p>
<p>                      A—B---C local/master</p>
<p>                  A—B---C origin/master</p>
<p>Now you <strong>edit some files in local</strong> master and create a new commit, at the same time someone else create a new commit to the origin master. The new history will be:</p>
<p>                      A—B---C—D local/master</p>
<p>                  A—B---C—E origin/master</p>
<p>You can see local/master and origin/master <strong>diverged</strong> because there are the commit D and E before the relative commit C. Now you want to try to pull the origin/master on local, but git cannot know <strong>what of the commits are made first</strong> so it cannot merge the origin/master into local/master. To prevent this you can create a <strong>new branch</strong> from the local/master before the new commits:</p>
<p>                                      local/new                   </p>
<p>                                    /</p>
<p>                      A—B---C local/master</p>
<p>                  A—B---C origin/master</p>
<p>Create the same commits as before:</p>
<p>                                      D local/new                   </p>
<p>                                    /</p>
<p>                      A—B---C local/master</p>
<p>                  A—B---C—E origin/master</p>
<p>Now you have 2 choices:</p>
<ol>
<li>
<strong>Pull the origin/master into local/master</strong> :</li>
</ol>
<p>                                      D local/new                   </p>
<p>                                    /</p>
<p>                      A—B---C—E local/master<br>
                                      /<br>
                  A—B---C—E origin/master</p>
<p><strong>Merge the local/new into local/master and push the local/master into origin</strong> :</p>
<p>                                       D       local/new</p>
<p>                                    /       \</p>
<p>                      A—B---C—E---D local/master<br>
                                      /      /<br>
                   A—B---C—E---D origin/master</p>
<ol start="2">
<li>
<strong>Push the local/new to origin/new</strong> :</li>
</ol>
<p>                                       D  local/new</p>
<p>                                    /      </p>
<p>                      A—B---C local/master<br>
                                          <br>
                   A—B---C—E origin/master</p>
<p>                                 \</p>
<p>                                   D origin/new</p>
<p>And from there (github) create a pull request to see if everything is ok and in the end merge (if possible, if not you can see <a href="https://github.community/t/pr-with-merge-conflicts-create-on-github/1688">this post</a>):</p>
<p>                  </p>
<p>                   A—B---C—E---D origin/master</p>
<p>                                 &amp;nbsp;     /</p>
<p>                                    D   origin/new</p>
<p>I know that it seems difficult but every time you do that become even more simply.</p>
<blockquote>
<p>Is there any way to do this? Would I ideally need to use a hook to do this?</p>
</blockquote>
<p>You can use the pre-push hook:</p>
<blockquote>
<p>The pre-push hook runs during git push, after the remote refs have been updated but before any objects have been transferred. It receives the name and location of the remote as parameters, and a list of to-be-updated refs through stdin. You can use it to validate a set of ref updates before a push occurs (a non-zero exit code will abort the push).</p>
</blockquote>
<p>but you dont need that, it’s better to use branch to do what you asked.</p>
<p>Hope this helps :slightly_smiling_face:  </p>
<p>-Gabriele-</p>
<blockquote>
<p><strong>Mark helpful posts with Accept as Solution to help other users locate important info. Don’t forget to give Kudos for great content!</strong></p>
</blockquote>
